Add theme style support to Learning Mode

yscik opened the pull request:
Fixes #4760

### Changes proposed in this Pull Request

- [x] Enable block templates independent of theme override
- [x] Add setting to enable theme styles. 
- [x] Skip redirect to `/learn` if theme styles are enabled
- [x] Don't override theme in editor if theme styles are enabled
- [x] Separate content styles not needed when theme styles are on
- [x] Review CSS for top themes and add `!important` everywhere
- [x] Add hooks to load custom CSS
- [x] WordPress 5.7 support
- [x] Add usage tracking

Other:
- Rename a few files and methods where different meanings of 'theme' was getting confusing

### Themes Checked

- [x] Divi
- [x] Astra
- [x] GeneratePress
- [x] Storefront
- [x] Hello Elementor
- [x] Twenty Twenty-One
- [x] OceanWP
- [x] Twenty Twenty
- [x] Twenty Twenty Two

### Testing instructions

- Open a lesson that has Learning Mode and make sure it still loads with the Sensei course theme
- Go to Sensei settings > Course > Learning Mode, check `Enable theme styles`
- Open a lesson that has LM. Should open without `/learn/` URL prefix, and use the active theme of the styles
- Check that things are not too broken in all the themes

<!-- Add the following only if there are new/updated actions or filters. Please provide a brief description of what they do and any arguments they may take. Be sure to also add the "Hooks" label to this PR. -->
### New/Updated Hooks

#### Filters:

* `sensei_course_learning_mode_theme_override_enabled` - Filters if the theme should be overriden for learning mode.
* `sensei_course_has_learning_mode_enabled` - Fixed the filtered value not being respected in all cases.

#### Actions:

* `sensei_course_learning_mode_load_theme` - Fires when learning mode is loaded for a page.
* `sensei_course_learning_mode_load_override_styles` - Fires when the override theme styles are loaded for learning mode.
* `sensei_course_learning_mode_override_theme` - Fires when the theme is override is added for learning mode.

<!-- Add the following only if there is any code that is being deprecated. Please list the replacement function or hook that should be called instead, if applicable. Be sure to also add the "Deprecation" label to this PR. -->
### Deprecated Code

* `Sensei_Course_Theme_Option::has_sensei_theme_enabled` renamed to `Sensei_Course_Theme_Option::has_learning_mode_enabled`
* `Sensei_Course_Theme_Option::should_use_sensei_theme` renamed to `Sensei_Course_Theme_Option::should_use_learning_mode`

<!--
Helpful tips for screenshots:
https://en.support.wordpress.com/make-a-screenshot/
-->
### Screenshot / Video

#### Theme styles

<img width="1560" alt="image" src="https://user-images.githubusercontent.com/176949/153087049-b01e7600-4449-4470-81e1-8374e8d88917.png">

#### Settings

<img width="697" alt="image" src="https://user-images.githubusercontent.com/176949/153045782-d18aef70-067d-4a16-aea6-719b211e867e.png">

yscik commented:
I'll look at adding some hooks and maybe tests, but otherwise it should be ready for review.

renatho reviewed (commented):
Looking good!

I'm a little concerned about future maintenance with the many `!important`, but if we didn't have a better approach to deal with all the themes, I think we have to live with that. 😜

And just some other things I noticed while testing it:

- I tested it with Twenty Twenty-One, and since we set the default background to `#FFF` in the body, it didn't get the green background.
- In the template editor, it didn't get the theme styles.
- When I activated the Twenty Twenty-Two theme, enabled the theme styles setting, and tried to edit the template through the lesson editor, it didn't load the theme blocks.
  <img width="1724" alt="Screen Shot 2022-02-10 at 16 53 26" src="https://user-images.githubusercontent.com/876340/153486408-ac8fe80b-824d-4380-81e4-945d441cc780.png">
- In the lesson sidebar, it didn't display the select with the template anymore.
  <img width="311" alt="Screen Shot 2022-02-10 at 16 36 01" src="https://user-images.githubusercontent.com/876340/153486656-ed1f8096-cc82-4097-8061-7b3c5a1f6429.png">

renatho commented:
Taking the opportunity that you're changing many styles in this PR, could you add a padding to the bottom of the quiz for resolutions >= 768px?

The reason is that the quiz bar is over the content (I have a "False" option after the "True") in the screenshot and I can't scroll to there.

<img width="1141" alt="Screen Shot 2022-02-10 at 17 58 16" src="https://user-images.githubusercontent.com/876340/153495793-1580fb36-1fb0-494d-85f4-beca7467de7d.png">

renatho commented:
> I tested it with Twenty Twenty-One, and since we set the default background to #FFF in the body, it didn't get the green background.

Actually, I think this item wouldn't be something very simple right? Because we need the background in different elements and we don't have a simple way to get that color. I thought about inherit, but it could become a problem too if it's a gradient. I'm thinking that maybe for the background would be better to keep the white as default as it is. Unless you know a good solution for that.

renatho commented on includes/course-theme/class-sensei-course-theme.php:
```diff
@@ -268,16 +303,29 @@ public function add_sensei_theme_body_class( $classes ) {
 	 * @access private
 	 */
 	public function enqueue_styles() {
-		Sensei()->assets->enqueue( self::THEME_NAME . '-style', 'css/sensei-course-theme.css' );
```
One more thing I realized while exploring the widget styles issue:
In the `Sensei_Course_Theme_Editor` we're still using the old names.

renatho replied:
After that, I also realized an alignment issue in the editor.

<img width="1477" alt="Screen Shot 2022-02-11 at 11 23 07" src="https://user-images.githubusercontent.com/876340/153608628-e3aea903-2f49-4f3e-b006-01dec90ee938.png">

yscik replied:
Worked on the styling issues in [5837a3b](https://github.com/Automattic/sensei/pull/4775/commits/5837a3bad5c7442853df4bfe500a5847df8940a8)

renatho replied:
I still could reproduce the issue.

Always in the editor, and in the frontend when I active the "Enable theme styles" setting. Tested with Twenty Twenty-Two.

<img width="986" alt="Screen Shot 2022-02-15 at 10 26 51" src="https://user-images.githubusercontent.com/876340/154071512-0124d0aa-ef44-4b6c-8684-da606c547a8b.png">

<img width="1475" alt="Screen Shot 2022-02-15 at 10 27 08" src="https://user-images.githubusercontent.com/876340/154071538-7149aa88-fbc3-4fab-b083-153edad5e214.png">

yscik commented:
> Theme background

Yeah, we can't really let it through in a way that looks good. But we can explore more when doing #4591.

> In the template editor, it didn't get the theme styles.

Note that in the site editor the theme styles only works for block themes. For others we have to override the theme, because the site editor won't load with a non-block theme.

> didn't load the theme blocks

Looks like this was because of the Gutenberg plugin, it still changes the site editor URL. Fixing in [67b4b43](https://github.com/Automattic/sensei/pull/4775/commits/67b4b433d6894a3e2a849ccaf47cf16e3ddb9065)

yscik commented on includes/course-theme/class-sensei-course-theme-option.php:
```diff
@@ -102,11 +103,22 @@ public function ensure_learning_mode_url_prefix() {
 	 * @return boolean
 	 */
 	public function should_use_sensei_theme() {
+		return self::should_use_learning_mode();
```
Good catch, added them in [67b4b43](https://github.com/Automattic/sensei/pull/4775/commits/67b4b433d6894a3e2a849ccaf47cf16e3ddb9065)

renatho replied:
Cool! Just one more thing that would be interesting, is to add the replacement function in the last argument. 🙂

yscik commented on includes/class-sensei-settings.php:
```diff
@@ -307,14 +307,23 @@ public function init_fields() {
 			'section'     => 'course-settings',
 		);
 
+		// Course Settings.
```
👍  [67b4b43](https://github.com/Automattic/sensei/pull/4775/commits/67b4b433d6894a3e2a849ccaf47cf16e3ddb9065)

yscik commented on includes/course-theme/class-sensei-course-theme-option.php:
```diff
@@ -132,31 +144,64 @@ public function should_use_sensei_theme() {
-		$enabled_via_filter = (bool) apply_filters( 'sensei_course_learning_mode_enabled', $enabled_for_course, $enabled_globally, $course_id );
-
-		return $enabled_for_course || $enabled_globally || $enabled_via_filter;
+		return (bool) apply_filters( 'sensei_course_learning_mode_enabled', $enabled, $course_id );
```
Since it'd be so complicated and confusing, and it's only been out for a little while, I'd say let's change the existing one. I added it to the updated hook list to be included in the release post, and if anyone is using it already they can easily update it or open an issue.

yscik commented:
* Removed the override from the lesson editor when theme styles are on. This also means the `Template` section and the whole template editing is not available this way for classic themes. (Might be a way to get this one working sometime.)
* Tried to make CSS a bit cleaner for setting a consistent content width and supporting alignments. Removed a few workarounds when it turned out that something is just broken for the theme even without learning mode.

renatho reviewed (approved):
Looks good and works well!

I just noticed some issues in smaller resolutions in the site editor. Since it's not perfect in the `master` too and probably it's not super critical, I'm already approving it and we can tweak it later. 😉

<img width="733" alt="Screen Shot 2022-02-16 at 15 56 11" src="https://user-images.githubusercontent.com/876340/154336543-626c0e37-426a-49d8-8e6b-1bee76be05c9.png">

<img width="780" alt="Screen Shot 2022-02-16 at 15 55 59" src="https://user-images.githubusercontent.com/876340/154336555-146139bf-4e2e-4709-b278-995be47015f8.png">

yscik commented:
Hm, I tested this for the last fix and it correctly adapts for me at the breakpoint, with the sidebar turning full-size too:

<img width="1298" alt="image" src="https://user-images.githubusercontent.com/176949/154368600-e3e5cb43-a947-4312-b8b3-75a95b1b7720.png">

renatho commented:
> Hm, I tested this for the last fix and it correctly adapts for me at the breakpoint, with the sidebar turning full-size too:

I checked that the issue is between 600px and 782px for me. Inspecting, I noticed this, where the variable is `0px`:

<img width="1181" alt="Screen Shot 2022-02-17 at 10 18 18" src="https://user-images.githubusercontent.com/876340/154489758-13432138-c199-41e3-9b46-17f510cf3e89.png">
